File: bash-server/clients/typescript/src/channels/command.ts
import { ServerError } from "../errors";
import { b64decode, makeMsg } from "../protocol";
import { CHAN_COMMAND, EvalResult, Message } from "../types";

type SendFn = (msg: Message) => Promise<void>;
type RecvFn = (ch: number, timeout?: number) => Promise<Message>;

export class CommandChannel {
  constructor(private send: SendFn, private recv: RecvFn) {}

  async eval(command: string, options?: { id?: string; timeout?: number }): Promise<EvalResult> {
    const msg = makeMsg(CHAN_COMMAND, "eval", { command });
    if (options?.id) (msg as any).id = options.id;
    await this.send(msg);

    let stdout = "";
    let stderr = "";
    let exit_code = -1;
    const timeout = options?.timeout ?? 30000;

    for (let i = 0; i < 3; i++) {
      const resp = await this.recv(CHAN_COMMAND, timeout);
      switch (resp.type) {
        case "stdout": {
          const data = String(resp.data ?? "");
          stdout = resp.encoding === "base64" ? b64decode(data) : data;
          break;
        }
        case "stderr": {
          const data = String(resp.data ?? "");
          stderr = resp.encoding === "base64" ? b64decode(data) : data;
          break;
        }
        case "complete":
          exit_code = Number(resp.exit_code ?? -1);
          break;
        case "error":
          throw new ServerError(String(resp.message ?? "eval error"), CHAN_COMMAND);
      }
    }

    return { stdout, stderr, exit_code };
  }
}

File: bash-server/clients/go/bashclient/command.go
package bashclient

import "context"

// CommandChannel handles channel 1: command evaluation.
type CommandChannel struct {
	send sendFunc
	recv recvFunc
}

func newCommandChannel(send sendFunc, recv recvFunc) *CommandChannel {
	return &CommandChannel{send: send, recv: recv}
}

// Eval evaluates a command and returns the result.
func (c *CommandChannel) Eval(ctx context.Context, command string) (*EvalResult, error) {
	msg := MakeMsg(ChanCommand, "eval", map[string]any{"command": command})
	if err := c.send(msg); err != nil {
		return nil, err
	}

	result := &EvalResult{ExitCode: -1}

	// Collect stdout, stderr, complete messages
	for i := 0; i < 3; i++ {
		resp, err := c.recv(ctx, ChanCommand)
		if err != nil {
			return nil, err
		}
		t := getString(resp, "type")
		switch t {
		case "stdout":
			data := getString(resp, "data")
			if getString(resp, "encoding") == "base64" {
				decoded, err := B64Decode(data)
				if err != nil {
					return nil, err
				}
				data = decoded
			}
			result.Stdout = data
		case "stderr":
			data := getString(resp, "data")
			if getString(resp, "encoding") == "base64" {
				decoded, err := B64Decode(data)
				if err != nil {
					return nil, err
				}
				data = decoded
			}
			result.Stderr = data
		case "complete":
			result.ExitCode = getInt(resp, "exit_code")
		case "error":
			return nil, newServerError(getString(resp, "message"), ChanCommand)
		}
	}

	return result, nil
}
